source/lac/trilinos_tpetra_sparsity_pattern.cc

```diff
@@ -0,0 +1,1078 @@
+                column_space_map,
+                graph,
+                nonlocal_graph);
+    }
+
+
+
+    template <typename Node>
+    void
+    SparsityPattern<Node>::reinit(
+      const IndexSet               &row_parallel_partitioning,
+      const IndexSet               &col_parallel_partitioning,
+      const MPI_Comm                communicator,
+      const std::vector<size_type> &n_entries_per_row)
+    {
+      SparsityPatternBase::resize(row_parallel_partitioning.size(),
+                                  col_parallel_partitioning.size());
+      Teuchos::RCP<MapType> row_map =
+        row_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
+      Teuchos::RCP<MapType> col_map =
+        col_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
+      reinit_sp(row_map,
+                col_map,
+                n_entries_per_row,
+                column_space_map,
+                graph,
+                nonlocal_graph);
+    }
+
+
+
+    template <typename Node>
+    void
+    SparsityPattern<Node>::reinit(const IndexSet &row_parallel_partitioning,
+                                  const IndexSet &col_parallel_partitioning,
+                                  const IndexSet &writable_rows,
+                                  const MPI_Comm  communicator,
+                                  const size_type n_entries_per_row)
+    {
+      SparsityPatternBase::resize(row_parallel_partitioning.size(),
+                                  col_parallel_partitioning.size());
+      Teuchos::RCP<MapType> row_map =
+        row_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
+      Teuchos::RCP<MapType> col_map =
+        col_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
+      reinit_sp(row_map,
+                col_map,
+                n_entries_per_row,
+                column_space_map,
+                graph,
+                nonlocal_graph);
+
+      IndexSet nonlocal_partitioner = writable_rows;
+      AssertDimension(nonlocal_partitioner.size(),
+                      row_parallel_partitioning.size());
+#  ifdef DEBUG
+      {
+        IndexSet tmp = writable_rows & row_parallel_partitioning;
+        Assert(tmp == row_parallel_partitioning,
+               ExcMessage(
+                 "The set of writable rows passed to this method does not "
+                 "contain the locally owned rows, which is not allowed."));
+      }
+#  endif
+      nonlocal_partitioner.subtract_set(row_parallel_partitioning);
+      if (Utilities::MPI::n_mpi_processes(communicator) > 1)
+        {
+          Teuchos::RCP<MapType> nonlocal_map =
+            nonlocal_partitioner.make_tpetra_map_rcp(communicator, true);
+          nonlocal_graph =
+            Utilities::Trilinos::internal::make_rcp<GraphType>(nonlocal_map,
+                                                               nonlocal_map,
+                                                               0);
```

Comment:
Is this correct? My understanding is that you are creating a sparsity pattern for the non-locally-owned rows. So the `row_map` needs to be `nonlocal_map`, but should't the column map still be the same column map used for the locally owned rows?

Reply:
I had to think about this carefully aswell.
But you should be right.

I changed the line 
```
          nonlocal_graph =
            Utilities::Trilinos::internal::make_rcp<GraphType>(nonlocal_map,
                                                               nonlocal_map,
                                                               0);
```
to
```
          nonlocal_graph =
            Utilities::Trilinos::internal::make_rcp<GraphType>(nonlocal_map,
                                                               col_map,
                                                               0);
```